odoo/models.py

```diff
@@ -1468,30 +1470,71 @@ def search_count(self, domain, limit=None):
                      list to match all records.
         :param limit: maximum number of record to count (upperbound) (default: all)
         """
-        res = self.search(domain, limit=limit, count=True)
-        return res if isinstance(res, int) else len(res)
+        query = self._search(domain, limit=limit)
+        return len(query)
 
     @api.model
-    @api.returns('self',
-        upgrade=lambda self, value, domain, offset=0, limit=None, order=None, count=False: value if count else self.browse(value),
-        downgrade=lambda self, value, domain, offset=0, limit=None, order=None, count=False: value if count else value.ids)
-    def search(self, domain, offset=0, limit=None, order=None, count=False):
-        """ search(domain[, offset=0][, limit=None][, order=None][, count=False])
+    @api.returns('self')
+    def search(self, domain, offset=0, limit=None, order=None):
+        """ search(domain[, offset=0][, limit=None][, order=None])
 
-        Searches for records based on the ``domain``
+        Search for the records that satisfy the given ``domain``
         :ref:`search domain <reference/orm/domains>`.
 
         :param domain: :ref:`A search domain <reference/orm/domains>`. Use an empty
                      list to match all records.
         :param int offset: number of results to ignore (default: none)
         :param int limit: maximum number of records to return (default: all)
         :param str order: sort string
-        :param bool count: if True, only counts and returns the number of matching records (default: False)
         :returns: at most ``limit`` records matching the search criteria
         :raise AccessError: if user is not allowed to access requested information
         """
-        res = self._search(domain, offset=offset, limit=limit, order=order, count=count)
-        return res if count else self.browse(res)
+        return self.search_fetch(domain, [], offset=offset, limit=limit, order=order)
+
+    @api.model
+    @api.returns('self')
+    def search_fetch(self, domain, field_names, offset=0, limit=None, order=None):
```

Comment:
```suggestion
    def _search_fetch(self, domain, field_names, offset=0, limit=None, order=None):
```

`search_fetch` should probably be a private method. As a public method it is a duplicate of `search_read`